dgraphtest/cluster.go

```diff
@@ -0,0 +1,301 @@
+	for i := 0; i < c.conf.numZeros; i++ {
+		if err := c.StartZero(i); err != nil {
+			return err
+		}
+	}
+	for i := 0; i < c.conf.numAlphas; i++ {
+		if err := c.StartAlpha(i); err != nil {
+			return err
+		}
+	}
+
+	if err := c.healthCheck(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Cluster) StartZero(id int) error {
+	if id >= c.conf.numZeros {
+		return fmt.Errorf("invalid id of zero: %v", id)
+	}
+	return c.startContainer(c.zeros[id])
+}
+
+func (c *Cluster) StartAlpha(id int) error {
+	if id >= c.conf.numAlphas {
+		return fmt.Errorf("invalid id of alpha: %v", id)
+	}
+	return c.startContainer(c.alphas[id])
+}
+
+func (c *Cluster) startContainer(dc dnode) error {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	if err := c.dcli.ContainerStart(ctx, dc.cid(), types.ContainerStartOptions{}); err != nil {
+		return errors.Wrapf(err, "error starting container [%v]", dc.cname())
+	}
+	return nil
+}
+
+func (c *Cluster) Stop() error {
+	c.log("stopping cluster with prefix [%v]", c.conf.prefix)
+	for i := range c.alphas {
+		if err := c.StopAlpha(i); err != nil {
+			return err
+		}
+	}
+	for i := range c.zeros {
+		if err := c.StopZero(i); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Cluster) StopZero(id int) error {
+	if id >= c.conf.numZeros {
+		return fmt.Errorf("invalid id of zero: %v", id)
+	}
+	return c.stopContainer(c.zeros[id])
+}
+
+func (c *Cluster) StopAlpha(id int) error {
+	if id >= c.conf.numAlphas {
+		return fmt.Errorf("invalid id of alpha: %v", id)
+	}
+	return c.stopContainer(c.alphas[id])
+}
+
+func (c *Cluster) stopContainer(dc dnode) error {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	if err := c.dcli.ContainerStop(ctx, dc.cid(), &stopTimeout); err != nil {
+		return errors.Wrapf(err, "error stopping container [%v]", dc.cname())
+	}
+	return nil
+}
+
+func (c *Cluster) Cleanup() {
+	c.log("cleaning up cluster with prefix [%v]", c.conf.prefix)
+
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+
+	ro := types.ContainerRemoveOptions{RemoveVolumes: true, Force: true}
+	for _, aa := range c.alphas {
+		if err := c.dcli.ContainerRemove(ctx, aa.cid(), ro); err != nil {
+			c.log("error removing alpha [%v]: %v", aa.cname(), err)
+		}
+	}
+	for _, zo := range c.zeros {
+		if err := c.dcli.ContainerRemove(ctx, zo.cid(), ro); err != nil {
+			c.log("error removing zero [%v]: %v", zo.cname(), err)
+		}
+	}
+	if err := c.dcli.NetworkRemove(ctx, c.net.id); err != nil {
+		c.log("error removing network [%v]: %v", c.net.name, err)
+	}
+}
+
+func (c *Cluster) createContainer(dc dnode) (string, error) {
+	cmd := dc.cmd(c)
+	image := c.dgraphImage()
+	mts, err := dc.mounts(c.conf)
+	if err != nil {
+		return "", err
+	}
+
+	cconf := &container.Config{Cmd: cmd, Image: image, WorkingDir: dc.workingDir(), ExposedPorts: dc.ports()}
```

Comment:
Can we log the final conf? Might be helpful while debugging.

Reply:
you could just do docker inspect or something. Not sure whether we can easily log all the config.